src/core/execute.c

```diff
@@ -3072,17 +3118,27 @@ int exec_spawn(Unit *unit,
                                fds, n_fds,
                                files_env,
                                unit->manager->user_lookup_fds[1],
-                               &exit_status);
+                               &exit_status,
+                               &error_message);
                 if (r < 0) {
                         log_open();
-                        log_struct_errno(LOG_ERR, r,
-                                         LOG_MESSAGE_ID(SD_MESSAGE_SPAWN_FAILED),
-                                         LOG_UNIT_ID(unit),
-                                         LOG_UNIT_MESSAGE(unit, "Failed at step %s spawning %s: %m",
-                                                          exit_status_to_string(exit_status, EXIT_STATUS_SYSTEMD),
-                                                          command->path),
-                                         "EXECUTABLE=%s", command->path,
-                                         NULL);
+                        if (error_message)
+                                log_struct_errno(LOG_ERR, r,
+                                                 LOG_MESSAGE_ID(SD_MESSAGE_SPAWN_FAILED),
+                                                 LOG_UNIT_ID(unit),
+                                                 LOG_UNIT_MESSAGE(unit, "%s: %m",
+                                                                  command->path),
```

Comment:
Pfff. /me wanders off to hide in the woods.

(Yeah, %m was supposed to be there. I didn't test the patch, obviously, but the final messages should look like this:

Failed to reset signal mask: %m
Failed to close remaining fds: %m
Execution cancelled: %m
Failed to update environment: %m
Failed to update dynamic user credentials: %m
UID validation failed for \""UID_FMT"\": %m
GID validation failed for \""GID_FMT"\": %m
Failed to determine user credentials: %m
Failed to determine group credentials: %m
Failed to determine supplementary groups: %m
Failed to send user credentials to PID1: %m
Failed to set up stdin: %m
Failed to set up stdout: %m
Failed to set up stderr: %m
Failed to attach to cgroup %s: %m
Failed to write /proc/self/oom_score_adj: %m
Failed to drop capabilities: %m
Failed to apply ambient capabilities (before UID change): %m
Failed to change UID to UID_FMT: %m
Failed to apply ambient capabilities (after UID change): %m
Failed to set SELinux context to %s: %m
Failed to set SMACK process label: %m
Failed to set secure bits: %m
Failed to disable new privileges: %m
Failed to restrict address families: %m
Failed to disable writing to executable memory: %m
Failed to apply realtime restrictions: %m
Failed to apply namespace restrictions: %m
Failed to apply sysctl restrictions: %m
Failed to apply module loading restrictions: %m
Failed to set up private devices: %m
Failed to apply syscall filters: %m
Failed to prepare process arguments: %m
Failed to prepare AppArmor profile change to %s: %m

I think this should be clearer than the previous message in all cases. The executable path is not included, because those errors are *before* the execution, so the binary should not matter.)